### PR TITLE
CA-261166: The jsonrpc version of the response should match the one o…

### PR DESCRIPTION
**Pull request by kc284**
…f the request. The structure of the error output for v2 should comply with the specs.

Signed-off-by: Konstantina Chremmou <konstantina.chremmou@citrix.com>

### CODE REVIEW

File: ocaml/xapi/api_server.ml [hunk 1]
```diff
@@ -123,8 +123,14 @@ let is_himn_req req =
...
 
+
+let json_of_error_object ?(data=None) code message =
+  let data_json = match data with Some d -> ["data", d] | None -> [] in
+  Rpc.Dict ([ "code", Rpc.Int code; "message", Rpc.String message; ] @ data_json)
+
+
 (* This bit is called directly by the fake_rpc callback *)
-let callback1 is_json req fd body call =
+let callback1 ?(json_ver=Jsonrpc.V1) is_json req fd body call =
```
Comment (mseri) [inline ocaml/xapi/api_server.ml]: I'd be verbose and use `json_version`

File: ocaml/xapi/api_server.ml [hunk 2]
```diff
@@ -196,11 +208,11 @@ let jsoncallback req bio _ =
   let fd = Buf_io.fd_of bio in (* fd only used for writing *)
   let body = Http_svr.read_body ~limit:Db_globs.http_limit_max_rpc_size req bio in
   try
-    let rpc = Jsonrpc.call_of_string body in
-    let response = Jsonrpc.a_of_response
+    let version, id, rpc = Jsonrpc.version_id_and_call_of_string body in
```
Comment (mseri) [inline ocaml/xapi/api_server.ml]: then here 

`let json_version, id, ...`

File: ocaml/xapi/api_server.ml [hunk 3]
```diff
@@ -196,11 +208,11 @@ let jsoncallback req bio _ =
...
+    let response = Jsonrpc.a_of_response ~id ~version
         ~empty:Bigbuffer.make
         ~append:(fun buf s -> Bigbuffer.append_substring buf s 0 (String.length s))
-        (callback1 false req fd (Some body) rpc) in
+        (callback1 ~json_ver:version true req fd (Some body) rpc) in
```
Comment (mseri) [inline ocaml/xapi/api_server.ml]: and here simply `callback ~json_version true ...`

File: ocaml/xapi/api_server.ml [hunk 4]
```diff
@@ -196,11 +208,11 @@ let jsoncallback req bio _ =
   let fd = Buf_io.fd_of bio in (* fd only used for writing *)
   let body = Http_svr.read_body ~limit:Db_globs.http_limit_max_rpc_size req bio in
   try
-    let rpc = Jsonrpc.call_of_string body in
-    let response = Jsonrpc.a_of_response
+    let version, id, rpc = Jsonrpc.version_id_and_call_of_string body in
+    let response = Jsonrpc.a_of_response ~id ~version
         ~empty:Bigbuffer.make
         ~append:(fun buf s -> Bigbuffer.append_substring buf s 0 (String.length s))
-        (callback1 false req fd (Some body) rpc) in
```
Comment (mseri) [inline ocaml/xapi/api_server.ml]: why were we using false?

Comment (kc284) [inline ocaml/xapi/api_server.ml]: No idea - copy-paste error?

File: ocaml/xapi/api_server.ml (hunk 3 above)
Comment (kc284) [inline ocaml/xapi/api_server.ml]: Then I need the label in line 212, but this is ok, better have the parameter of callback1 named properly.

**Pull request merged**